qcfractal/storage_sockets/models.py

```diff
@@ -0,0 +1,259 @@
+
+    meta = {
+        'collection': 'collections',  # DB collection/table name
+        'indexes': [
+            {'fields': ('collection_type', 'name'), 'unique': True}
+        ]
+    }
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+class Molecule(db.DynamicDocument):
+    """
+        The molecule DB collection is managed by pymongo, so far
+    """
+
+    name = db.StringField()
+    molecular_formula = db.StringField()
+    molecule_hash = db.StringField()
+
+    def create_hash(self):
+        """ TODO: create a special hash before saving"""
+        return ''
+
+    def save(self, *args, **kwargs):
+        """Override save to add molecule_hash"""
+        # self.molecule_hash = self.create_hash()
+
+        return super(Molecule, self).save(*args, **kwargs)
+
+    meta = {
+        'collection': 'molecules',
+        'indexes': [
+            {'fields': ('molecule_hash', 'molecular_formula'), 'unique': True}
+        ]
+    }
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+class Options(db.DynamicDocument):
+    """
+        Options are unique for a specific program and name
+    """
+
+    program = db.StringField(required=True)
+    # TODO: choose a more descriptive name, like options_type
+    option_name = db.StringField(default='default', required=True)
+
+    # program specific
+    # e_convergence = db.IntField()
+    # d_convergence = db.IntField()
+    # dft_spherical_points = db.IntField()
+    # dft_radial_points = db.IntField()
+    # maxiter = db.IntField()
+    # scf_type = db.StringField()
+    # mp2_type = db.StringField()
+    # freeze_core = db.BooleanField()
+
+    meta = {
+        'indexes': [
+            {'fields': ('program', 'option_name'), 'unique': True}
+        ]
+    }
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+class BaseResult(db.DynamicDocument):
+    """
+        Abstract Base class for Results and Procedures
+    """
+
+    # queue related
+    task_queue_id = db.StringField()  # ObjectId, reference task_queue but without validation
+    status = db.StringField(default='INCOMPLETE', choices=['COMPLETE, INCOMPLETE', 'ERROR'])
+
+    meta = {
+        'abstract': True,
+        # 'allow_inheritance': True,
+        'indexes': [
+            'status'
+        ]
+    }
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+class Result(BaseResult):
+    """
+        Hold the result of an atomic single calculation
+    """
+
+    # uniquely identifying a result
+    program = db.StringField(required=True)  # example "rdkit", is it the same as program in options?
+    driver = db.StringField(required=True)  # example "gradient"
+    method = db.StringField(required=True)  # example "uff"
+    basis = db.StringField()
+    molecule = db.ReferenceField(Molecule, required=True)   # or LazyReferenceField if only ID is needed?
```

Comment:
Only ID is usually needed as we typically only pull data from `return_result` or `properties`.

Reply:
Which field are you referring to?

By referencing the class we mean it's going to be pulled and used. If we use class_id only, then it means we will rarely use it or pull it. 
So